src/ws/cockpitauthpipe.c

```diff
@@ -0,0 +1,466 @@
+
+  g_signal_emit (self, signals[MESSAGE], 0, data);
+
+  if (self->timeout)
+    g_source_remove (self->timeout);
+  self->timeout = 0;
+
+  if (!self->pipe_closed)
+    {
+      self->timeout = g_timeout_add_seconds (self->max_idle,
+                                             on_timeout, self);
+    }
+}
+
+static void
+on_pipe_read (CockpitPipe *pipe,
+              GByteArray *input,
+              gboolean end_of_data,
+              gpointer user_data)
+{
+  CockpitAuthPipe *self = user_data;
+  GBytes *message = NULL;
+
+  /* Let close report the result */
+  if (end_of_data)
+    return;
+
+  /* We expect every read to be a complete message */
+  if (input->len > 0)
+    {
+      message = cockpit_pipe_consume (input, 0, input->len, 0);
+      report_message (self, message);
+      g_bytes_unref (message);
+    }
+}
+
+static void
+on_pipe_close (CockpitPipe *pipe,
+               const gchar *problem,
+               gpointer user_data)
+{
+  CockpitAuthPipe *self = user_data;
+  GByteArray *buf = NULL;
+  GBytes *message = NULL;
+  GError *error = NULL;
+
+  if (self->pipe_closed)
+    return;
+
+  self->pipe_closed = TRUE;
+
+  if (self->timeout)
+    g_source_remove (self->timeout);
+  self->timeout = 0;
+
+  buf = cockpit_pipe_get_buffer (pipe);
+
+  if (problem)
+      g_warning ("%s: Auth pipe closed: %s", self->logname, problem);
+  else
+      g_debug ("%s: Auth pipe closed", self->logname);
+
+  if (problem && g_strcmp0 (problem, "timeout") == 0)
+    {
+      g_set_error (&error, COCKPIT_ERROR, COCKPIT_ERROR_FAILED,
+                   "%s", "Authentication failed: Timeout");
+    }
+  else if (problem)
+    {
+      g_set_error (&error, COCKPIT_ERROR, COCKPIT_ERROR_FAILED,
+                   "%s", "Internal error in login process");
+    }
+  else if (self->send_signal && !message)
+    {
+      g_set_error (&error, COCKPIT_ERROR, COCKPIT_ERROR_FAILED,
+                   "%s", "Authentication failed: no results");
+    }
+
+  if (buf->len)
+    {
+      message = cockpit_pipe_consume (buf, 0, buf->len, 0);
+      report_message (self, message);
+      g_bytes_unref (message);
+    }
+
+  g_signal_emit (self, signals[CLOSED], 0, error);
+
+  if (error)
+    g_error_free (error);
+
+  cockpit_auth_pipe_close (self, NULL);
+
+  // This should only run once, release the refs we took when attaching the signal
+  g_object_unref (pipe);
+  g_object_unref (self);
+}
+
+static void
+cockpit_auth_pipe_init (CockpitAuthPipe *self)
+{
+  self->max_idle = cockpit_ws_auth_pipe_timeout;
+  self->max_wait_pipe = cockpit_ws_auth_pipe_timeout;
+  self->send_signal = FALSE;
+  self->id = NULL;
+}
+
+static void
+cockpit_auth_pipe_constructed (GObject *object)
+{
+  int pair[2];
+
+  CockpitAuthPipe *self = COCKPIT_AUTH_PIPE (object);
+
+  G_OBJECT_CLASS (cockpit_auth_pipe_parent_class)->constructed (object);
+
+  if (socketpair (AF_UNIX, SOCK_SEQPACKET, 0, pair) < 0)
+    {
+      g_error ("%s: Couldn't create socket pair: %s",
```

Comment:
EMFILE or ENFILE seem like errors that shouldn't kill the process? Or did you already consider them?